### ISSUE TITLE
Reduced size in pem_password_cb

### ISSUE THREAD

**Issue by mezen** [openssl/openssl#16601]
OpenSSL 3.0.0 calls `pem_password_cb` with `size = 1022`, but OpenSSL 1.1.1 calls it with `size = 1024`.

In my case the passphrase has a length of 1023, last byte for null terminator. That works with OpenSSL 1.1.1 great, but with OpenSSL 3.0.0 it fails always, because I am unable to use the correct passphrase :(

**Comment by levitte** [comment]
Uhmm, password callbacks of type `pem_password_cb` is called all over the place.  Can you give us a precise example?

**Comment by mezen** [comment]
I try to setup a TLS server and the call `SSL_CTX_use_PrivateKey_file` triggers the callback `pem_password_cb` which was set with `SSL_CTX_set_default_passwd_cb`.

I cannot produce any small snippet or debug OpenSSL, I am using OpenSSL on Windows as DLL from another language (Delphi).

**Comment by mezen** [comment]
I was able to write a simple test application for your help. It is written in Delphi, but should still be readable:

```
function GetPasswordCallback(buf: PIdAnsiChar; size: TIdC_INT; rwflag: TIdC_INT; userdata: Pointer): TIdC_INT; cdecl;
begin
  Writeln('size: ', size);
  Result := 0;
end;

procedure Main;
var
  LCtx: PSSL_CTX;
begin
  LCtx := SSL_CTX_new(TLS_method());
  SSL_CTX_set_mode(LCtx, SSL_MODE_AUTO_RETRY);
  SSL_CTX_set_default_verify_paths(LCtx);
  SSL_CTX_set_session_cache_mode(LCtx, SSL_SESS_CACHE_BOTH);
  SSL_CTX_set_default_passwd_cb(LCtx, GetPasswordCallback);

  SSL_CTX_use_certificate_chain_file(LCtx, GetPAnsiChar('C:\server.crt'));
  SSL_CTX_use_PrivateKey_file(LCtx, GetPAnsiChar('C:\server.pem'), SSL_FILETYPE_PEM);
  SSL_CTX_check_private_key(LCtx);
end;
```
`GetPasswordCallback` will be called 2 times. First with `size = 1022`, second time with `size = 1024`.

**Comment by levitte** [comment]
Thanks, I translated that into C, and saw the same results.

``` C
#include <openssl/core.h>
#include <openssl/types.h>
#include <openssl/ssl.h>

pem_password_cb GetPasswordCallback;
int GetPasswordCallback(char *buf, int size, int rw_flag, void *userdata)
{
  printf("size: %d\n", size);
  return 0;
}

int main(int argc, char *argv[])
{
  SSL_CTX *LCtx = SSL_CTX_new(TLS_method());
  SSL_CTX_set_mode(LCtx, SSL_MODE_AUTO_RETRY);
  SSL_CTX_set_default_verify_paths(LCtx);
  SSL_CTX_set_session_cache_mode(LCtx, SSL_SESS_CACHE_BOTH);
  SSL_CTX_set_default_passwd_cb(LCtx, GetPasswordCallback);

  SSL_CTX_use_certificate_chain_file(LCtx, "server.crt");
  SSL_CTX_use_PrivateKey_file(LCtx, "server.pem", SSL_FILETYPE_PEM);
  SSL_CTX_check_private_key(LCtx);
}
```

I'll have a closer look tomorrow

**Comment by mezen** [comment]
Since 20 days have already passed: Is there any progress yet? Unfortunately this problem blocks an upgrade to 3.0.0.

**Comment by kaduk** [comment]
ping @levitte ; any updates?

**Comment by mezen** [comment]
Any updates? Unfortunately this problem still blocks an upgrade to 3.0.0 :(

**Comment by kaduk** [comment]
@levitte this is still assigned to you -- are you still working on it or should it be returned to the pool?

**Comment by t8m** [comment]
I'll try to look at it.

**Comment by t8m** [comment]
Fix in #17320 - hopefully it is correct.
